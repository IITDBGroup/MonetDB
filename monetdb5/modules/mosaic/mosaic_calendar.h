/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2018 MonetDB B.V.
 */

/*
 * (c)2014 author Martin Kersten
 */

#ifndef _MOSAIC_TEMPORAL_
#define _MOSAIC_TEMPORAL_

#include <mal.h>
#include "mal_interpreter.h"
#include "mal_client.h"

#define TEMPORALTHRESHOLD 4192
#define TEMPORALSIZE 256 

bool MOStypes_calendar(BAT* b);
mal_export void MOScreatecalendar(Client cntxt, MOStask task);
mal_export void MOSlayout_calendar(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSlayout_calendar_hdr(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSadvance_calendar(Client cntxt, MOStask task);
mal_export void MOSskip_calendar(Client cntxt, MOStask task);
mal_export flt  MOSestimate_calendar(Client cntxt, MOStask task);
mal_export void MOScompress_calendar(Client cntxt, MOStask task);
mal_export void MOSdecompress_calendar(Client cntxt, MOStask task);
mal_export str MOSselect_calendar(Client cntxt,  MOStask task, void *low, void *hgh, bit *li, bit *hi, bit *anti);
mal_export str MOSthetaselect_calendar(Client cntxt,  MOStask task, void *val, str oper);
mal_export str MOSprojection_calendar(Client cntxt,  MOStask task);
mal_export str MOSjoin_calendar(Client cntxt,  MOStask task);
#endif /* _MOSAIC_TEMPORAL_ */
