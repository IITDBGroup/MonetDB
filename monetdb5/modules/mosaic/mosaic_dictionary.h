/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2018 MonetDB B.V.
 */

/*
 * (c)2014 author Martin Kersten
 */

#ifndef _MOSAIC_DICT_
#define _MOSAIC_DICT_

#include <mal.h>
#include "mal_interpreter.h"
#include "mal_client.h"

#define DICTTHRESHOLD 4192
#define DICTSIZE 256 

mal_export void MOScreatedictionary(Client cntxt, MOStask task);
mal_export void MOSlayout_dictionary(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSlayout_dictionary_hdr(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSadvance_dictionary(Client cntxt, MOStask task);
mal_export void MOSskip_dictionary(Client cntxt, MOStask task);
mal_export flt  MOSestimate_dictionary(Client cntxt, MOStask task);
mal_export void MOScompress_dictionary(Client cntxt, MOStask task);
mal_export void MOSdecompress_dictionary(Client cntxt, MOStask task);
mal_export str MOSselect_dictionary(Client cntxt,  MOStask task, void *low, void *hgh, bit *li, bit *hi, bit *anti);
mal_export str MOSthetaselect_dictionary(Client cntxt,  MOStask task, void *val, str oper);
mal_export str MOSprojection_dictionary(Client cntxt,  MOStask task);
mal_export str MOSjoin_dictionary(Client cntxt,  MOStask task);
#endif /* _MOSAIC_DICT_ */
