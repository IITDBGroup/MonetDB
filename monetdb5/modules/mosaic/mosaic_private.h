#define PREVVALUEbit(x) ((x) - 1)
#define PREVVALUEbte(x) ((x) - 1)
#define PREVVALUEsht(x) ((x) - 1)
#define PREVVALUEint(x) ((x) - 1)
#define PREVVALUElng(x) ((x) - 1)
#define PREVVALUEoid(x) ((x) - 1)
#ifdef HAVE_HGE
#define PREVVALUEhge(x) ((x) - 1)
#endif
#define PREVVALUEflt(x) nextafterf((x), -GDK_flt_max)
#define PREVVALUEdbl(x) nextafter((x), -GDK_dbl_max)
#define PREVVALUEdate(x) ((x) - 1)
#define PREVVALUEdaytime(x) ((x) - 1)

#define NEXTVALUEbit(x) ((x) + 1)
#define NEXTVALUEbte(x) ((x) + 1)
#define NEXTVALUEsht(x) ((x) + 1)
#define NEXTVALUEint(x) ((x) + 1)
#define NEXTVALUElng(x) ((x) + 1)
#define NEXTVALUEoid(x) ((x) + 1)
#ifdef HAVE_HGE
#define NEXTVALUEhge(x) ((x) + 1)
#endif
#define NEXTVALUEflt(x) nextafterf((x), GDK_flt_max)
#define NEXTVALUEdbl(x) nextafter((x), GDK_dbl_max)
#define NEXTVALUEdate(x) ((x) + 1)
#define NEXTVALUEdaytime(x) ((x) + 1)
