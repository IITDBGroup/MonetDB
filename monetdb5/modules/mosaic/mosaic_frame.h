/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2018 MonetDB B.V.
 */

/*
 * (c)2014 author Martin Kersten
 */

#ifndef _MOSAIC_FRAME_
#define _MOSAIC_FRAME_

#include <mal.h>
#include "mal_interpreter.h"
#include "mal_client.h"
#include "mosaic_dictionary.h"

bool MOStypes_frame(BAT* b);
mal_export void MOScreateframeDictionary(Client cntxt, MOStask task);
mal_export void MOSlayout_frame_hdr(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSlayout_frame(Client cntxt, MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSadvance_frame(Client cntxt, MOStask task);
mal_export void MOSskip_frame(Client cntxt, MOStask task);
mal_export flt  MOSestimate_frame(Client cntxt, MOStask task);
mal_export void MOScompress_frame(Client cntxt, MOStask task);
mal_export void MOSdecompress_frame(Client cntxt, MOStask task);
mal_export str MOSselect_frame(Client cntxt,  MOStask task, void *low, void *hgh, bit *li, bit *hi, bit *anti);
mal_export str MOSthetaselect_frame(Client cntxt,  MOStask task, void *val, str oper);
mal_export str MOSprojection_frame(Client cntxt,  MOStask task);
mal_export str MOSjoin_frame(Client cntxt,  MOStask task);
#endif /* _MOSAIC_FRAME_ */
