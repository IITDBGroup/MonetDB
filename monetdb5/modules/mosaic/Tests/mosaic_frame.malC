b:= bat.new(:timestamp);
bat.append(b,"2014-08-14 12:00:00":timestamp);
bat.append(b,"2014-08-14 12:00:00":timestamp);
bat.append(b,"2014-08-14 12:01:00":timestamp);
bat.append(b,"2014-08-14 12:02:00":timestamp);
bat.append(b,"2014-08-14 12:00:00":timestamp);
bat.append(b,"2014-08-14 12:00:00":timestamp);
bat.append(b,"2014-08-14 12:01:00":timestamp);
bat.append(b,"2014-08-14 12:02:00":timestamp);
bat.append(b,"2014-08-14 12:00:00":timestamp);
b:= bat.append(b,b);
base:= algebra.copy(b);

mosaic.compress(b,"frame");
b:= mosaic.decompress(b);
io.print(base,b);

d:= bat.new(:date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-15":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-14":date);
bat.append(d,"2014-08-15":date);
d:= bat.append(d,d);
dbase:= algebra.copy(d);

mosaic.compress(d,"frame");
d:= mosaic.decompress(d);
io.print(dbase,d);

e:= bat.new(:date);
bat.append(e,"1996-03-13":date);
bat.append(e,"1996-04-12":date);
bat.append(e,"1996-01-29":date);
bat.append(e,"1996-04-21":date);
bat.append(e,"1996-03-30":date);
bat.append(e,"1996-01-30":date);
bat.append(e,"1997-01-28":date);
bat.append(e,"1994-02-02":date);
bat.append(e,"1993-11-09":date);
bat.append(e,"1994-01-16":date);
bat.append(e,"1993-12-04":date);
bat.append(e,"1993-12-14":date);
bat.append(e,"1993-10-29":date);
bat.append(e,"1996-01-10":date);
bat.append(e,"1994-10-31":date);
bat.append(e,"1994-10-16":date);
bat.append(e,"1994-08-08":date);
bat.append(e,"1992-04-27":date);
bat.append(e,"1996-05-07":date);
bat.append(e,"1996-02-01":date);
bat.append(e,"1996-01-15":date);
bat.append(e,"1996-03-21":date);
bat.append(e,"1996-02-11":date);
bat.append(e,"1996-01-16":date);
bat.append(e,"1996-02-10":date);
bat.append(e,"1995-10-23":date);
bat.append(e,"1995-08-14":date);
bat.append(e,"1995-08-07":date);
bat.append(e,"1995-08-04":date);
bat.append(e,"1995-08-28":date);
bat.append(e,"1995-07-21":date);
bat.append(e,"1993-10-29":date);
bat.append(e,"1993-12-09":date);
bat.append(e,"1993-12-09":date);
bat.append(e,"1993-11-09":date);
bat.append(e,"1998-10-23":date);
bat.append(e,"1998-10-09":date);
bat.append(e,"1998-10-30":date);
bat.append(e,"1996-02-21":date);
bat.append(e,"1996-01-22":date);
bat.append(e,"1996-01-19":date);
bat.append(e,"1995-11-26":date);
bat.append(e,"1995-11-08":date);
bat.append(e,"1996-02-01":date);
bat.append(e,"1996-02-03":date);
bat.append(e,"1992-07-21":date);
bat.append(e,"1992-07-02":date);
bat.append(e,"1992-07-10":date);
bat.append(e,"1996-09-29":date);
bat.append(e,"1996-11-14":date);
bat.append(e,"1996-11-04":date);
bat.append(e,"1996-09-26":date);
bat.append(e,"1996-10-02":date);
bat.append(e,"1996-10-17":date);
bat.append(e,"1996-12-08":date);
bat.append(e,"1994-09-30":date);
bat.append(e,"1995-04-20":date);
bat.append(e,"1995-07-17":date);
bat.append(e,"1995-07-06":date);
bat.append(e,"1994-02-19":date);
bat.append(e,"1994-02-21":date);
bat.append(e,"1997-04-17":date);
bat.append(e,"1997-01-27":date);
bat.append(e,"1997-02-20":date);
bat.append(e,"1997-03-18":date);
bat.append(e,"1997-04-19":date);
bat.append(e,"1997-01-25":date);
bat.append(e,"1998-07-04":date);
bat.append(e,"1998-06-26":date);
bat.append(e,"1998-08-13":date);
bat.append(e,"1998-06-27":date);
bat.append(e,"1998-06-19":date);
bat.append(e,"1998-08-11":date);
bat.append(e,"1998-06-24":date);
bat.append(e,"1994-08-17":date);
bat.append(e,"1994-08-24":date);
bat.append(e,"1994-07-02":date);
bat.append(e,"1994-06-06":date);
bat.append(e,"1994-07-31":date);
bat.append(e,"1994-10-03":date);
bat.append(e,"1994-01-12":date);
bat.append(e,"1994-03-03":date);
bat.append(e,"1994-01-26":date);
bat.append(e,"1994-03-17":date);
bat.append(e,"1994-02-13":date);
bat.append(e,"1994-01-26":date);
bat.append(e,"1998-04-10":date);
bat.append(e,"1998-05-23":date);
bat.append(e,"1998-02-23":date);
bat.append(e,"1998-04-12":date);
bat.append(e,"1998-01-29":date);
bat.append(e,"1998-03-05":date);
bat.append(e,"1994-07-19":date);
bat.append(e,"1994-06-03":date);
bat.append(e,"1993-04-01":date);
bat.append(e,"1993-04-13":date);
bat.append(e,"1993-05-14":date);
bat.append(e,"1994-12-24":date);
bat.append(e,"1994-12-01":date);
bat.append(e,"1994-12-30":date);

ebase:= algebra.copy(e);
mosaic.compress(e,"frame");
e:= mosaic.decompress(e);
io.print(ebase,e);
#mosaic.analyse("frame prefix");
