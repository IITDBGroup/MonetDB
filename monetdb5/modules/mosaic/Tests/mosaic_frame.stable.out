stdout of test 'mosaic_frame` in directory 'monetdb5/modules/mosaic` itself:


# 07:42:13 >  
# 07:42:13 >  "mserver5" "--debug=10" "--set" "gdk_nr_threads=0" "--set" "mapi_open=true" "--set" "mapi_port=30639" "--set" "mapi_usock=/var/tmp/mtest-1108/.s.monetdb.30639" "--set" "monet_prompt=" "--forcemito" "--set" "mal_listing=2" "--dbpath=/export/scratch1/mk/mosaic//Linux/var/MonetDB/mTests_monetdb5_modules_mal" "mosaic_frame.mal"
# 07:42:13 >  

# MonetDB 5 server v11.20.0
# This is an unreleased version
# Serving database 'mTests_monetdb5_modules_mal', using 8 threads
# Compiled for x86_64-unknown-linux-gnu/64bit with 64bit OIDs and 128bit integers dynamically linked
# Found 15.590 GiB available main-memory.
# Copyright (c) 1993-July 2008 CWI.
# Copyright (c) August 2008-2014 MonetDB B.V., all rights reserved
# Visit http://www.monetdb.org/ for further information
# Listening for connection requests on mapi:monetdb://vienna.ins.cwi.nl:30639/
# Listening for UNIX domain connection requests on mapi:monetdb:///var/tmp/mtest-1108/.s.monetdb.30639
# MonetDB/GIS module loaded
# MonetDB/SQL module loaded
# MonetDB/R   module loaded

# 11:50:49 >  
# 11:50:49 >  "mclient" "-lmal" "-ftest" "-tnone" "-Eutf-8" "--host=/var/tmp/mtest-30332" "--port=39586"
# 11:50:49 >  

# 15:01:28 >  
# 15:01:28 >  "mclient" "-lmal" "-ftest" "-Eutf-8" "-i" "-e" "--host=/var/tmp/mtest-15809" "--port=37660"
# 15:01:28 >  

#x:= mosaic.compress(b,"frame");
#bits 2[0] 3160340058811904 [1] 3160340058871904 [2] 3160340058931904 
#io.print(base,z);
#--------------------------#
# h	t  # name
# void	timestamp  # type
#--------------------------#
[ 0@0,	2014-08-14 12:00:00.000,	2014-08-14 12:00:00.000  ]
[ 1@0,	2014-08-14 12:00:00.000,	2014-08-14 12:00:00.000  ]
[ 2@0,	2014-08-14 12:01:00.000,	2014-08-14 12:01:00.000  ]
[ 3@0,	2014-08-14 12:02:00.000,	2014-08-14 12:02:00.000  ]
[ 4@0,	2014-08-14 12:00:00.000,	2014-08-14 12:00:00.000  ]
[ 5@0,	2014-08-14 12:00:00.000,	2014-08-14 12:00:00.000  ]
[ 6@0,	2014-08-14 12:00:00.000,	2014-08-14 12:00:00.000  ]
[ 7@0,	2014-08-14 12:01:00.000,	2014-08-14 12:01:00.000  ]
[ 8@0,	2014-08-14 12:02:00.000,	2014-08-14 12:02:00.000  ]
[ 9@0,	2014-08-14 12:00:00.000,	2014-08-14 12:00:00.000  ]
# framebits 1:[0] 0 [1] 1 
# clk 11	sizes         40	     12608	31520%	     0.003x	frame[1,10]	
#--------------------------#
# t	t	t  # name
# void	date	date  # type
#--------------------------#
[ 0@0,	2014-08-14,	2014-08-14  ]
[ 1@0,	2014-08-14,	2014-08-14  ]
[ 2@0,	2014-08-14,	2014-08-14  ]
[ 3@0,	2014-08-14,	2014-08-14  ]
[ 4@0,	2014-08-15,	2014-08-15  ]
[ 5@0,	2014-08-14,	2014-08-14  ]
[ 6@0,	2014-08-14,	2014-08-14  ]
[ 7@0,	2014-08-14,	2014-08-14  ]
[ 8@0,	2014-08-14,	2014-08-14  ]
[ 9@0,	2014-08-15,	2014-08-15  ]
# framebits 7:[0] -1416 [1] -1350 [2] -1342 [3] -1331 [4] -1077 [5] -1065 [6] -1034 [7] -866 [8] -855 [9] -830 [10] -825 [11] -820 [12] -791 [13] -787 [14] -777 [15] -770 [16] -759 [17] -753 [18] -751 [19] -741 [20] -727 [21] -649 [22] -646 [23] -620 [24] -603 [25] -591 [26] -583 [27] -574 [28] -567 [29] -530 [30] -527 [31] -514 [32] -499 [33] -468 [34] -445 [35] -439 [36] -328 [37] -251 [38] -240 [39] -236 [40] -222 [41] -219 [42] -212 [43] -198 [44] -142 [45] -126 [46] -108 [47] -63 [48] -58 [49] -57 [50] -54 [51] -51 [52] -44 [53] -43 [54] -41 [55] -39 [56] -32 [57] -31 [58] -21 [59] 0 [60] 8 [61] 17 [62] 30 [63] 39 [64] 55 [65] 197 [66] 200 [67] 203 [68] 218 [69] 236 [70] 246 [71] 270 [72] 318 [73] 320 [74] 321 [75] 344 [76] 370 [77] 400 [78] 402 [79] 687 [80] 712 [81] 722 [82] 758 [83] 760 [84] 801 [85] 828 [86] 833 [87] 835 [88] 836 [89] 843 [90] 881 [91] 883 [92] 940 [93] 954 [94] 961 
# clk 119	sizes        400	     12688	3172%	     0.032x	frame[1,100]	
# framebits 7:[0] -1416 [1] -1350 [2] -1342 [3] -1331 [4] -1077 [5] -1065 [6] -1034 [7] -866 [8] -855 [9] -830 [10] -825 [11] -820 [12] -791 [13] -787 [14] -777 [15] -770 [16] -759 [17] -753 [18] -751 [19] -741 [20] -727 [21] -649 [22] -646 [23] -620 [24] -603 [25] -591 [26] -583 [27] -574 [28] -567 [29] -530 [30] -527 [31] -514 [32] -499 [33] -468 [34] -445 [35] -439 [36] -328 [37] -251 [38] -240 [39] -236 [40] -222 [41] -219 [42] -212 [43] -198 [44] -142 [45] -126 [46] -108 [47] -63 [48] -58 [49] -57 [50] -54 [51] -51 [52] -44 [53] -43 [54] -41 [55] -39 [56] -32 [57] -31 [58] -21 [59] 0 [60] 8 [61] 17 [62] 30 [63] 39 [64] 55 [65] 197 [66] 200 [67] 203 [68] 218 [69] 236 [70] 246 [71] 270 [72] 318 [73] 320 [74] 321 [75] 344 [76] 370 [77] 400 [78] 402 [79] 687 [80] 712 [81] 722 [82] 758 [83] 760 [84] 801 [85] 828 [86] 833 [87] 835 [88] 836 [89] 843 [90] 881 [91] 883 [92] 940 [93] 954 [94] 961 
#--------------------------#
# t	t	t  # name
# void	date	date  # type
#--------------------------#
[ 0@0,	1996-03-13,	1996-03-13  ]
[ 1@0,	1996-04-12,	1996-04-12  ]
[ 2@0,	1996-01-29,	1996-01-29  ]
[ 3@0,	1996-04-21,	1996-04-21  ]
[ 4@0,	1996-03-30,	1996-03-30	]
[ 5@0,	1996-01-30,	1996-01-30  ]
[ 6@0,	1997-01-28,	1997-01-28  ]
[ 7@0,	1994-02-02,	1994-02-02  ]
[ 8@0,	1993-11-09,	1993-11-09  ]
[ 9@0,	1994-01-16,	1994-01-16  ]
[ 10@0,	1993-12-04,	1993-12-04  ]
[ 11@0,	1993-12-14,	1993-12-14  ]
[ 12@0,	1993-10-29,	1993-10-29  ]
[ 13@0,	1996-01-10,	1996-01-10  ]
[ 14@0,	1994-10-31,	1994-10-31  ]
[ 15@0,	1994-10-16,	1994-10-16  ]
[ 16@0,	1994-08-08,	1994-08-08  ]
[ 17@0,	1992-04-27,	1992-04-27  ]
[ 18@0,	1996-05-07,	1996-05-07	]
[ 19@0,	1996-02-01,	1996-02-01  ]
[ 20@0,	1996-01-15,	1996-01-15  ]
[ 21@0,	1996-03-21,	1996-03-21  ]
[ 22@0,	1996-02-11,	1996-02-11  ]
[ 23@0,	1996-01-16,	1996-01-16  ]
[ 24@0,	1996-02-10,	1996-02-10  ]
[ 25@0,	1995-10-23,	1995-10-23  ]
[ 26@0,	1995-08-14,	1995-08-14  ]
[ 27@0,	1995-08-07,	1995-08-07	]
[ 28@0,	1995-08-04,	1995-08-04  ]
[ 29@0,	1995-08-28,	1995-08-28  ]
[ 30@0,	1995-07-21,	1995-07-21  ]
[ 31@0,	1993-10-29,	1993-10-29  ]
[ 32@0,	1993-12-09,	1993-12-09  ]
[ 33@0,	1993-12-09,	1993-12-09  ]
[ 34@0,	1993-11-09,	1993-11-09  ]
[ 35@0,	1998-10-23,	1998-10-23  ]
[ 36@0,	1998-10-09,	1998-10-09	]
[ 37@0,	1998-10-30,	1998-10-30  ]
[ 38@0,	1996-02-21,	1996-02-21  ]
[ 39@0,	1996-01-22,	1996-01-22  ]
[ 40@0,	1996-01-19,	1996-01-19  ]
[ 41@0,	1995-11-26,	1995-11-26	]
[ 42@0,	1995-11-08,	1995-11-08  ]
[ 43@0,	1996-02-01,	1996-02-01  ]
[ 44@0,	1996-02-03,	1996-02-03  ]
[ 45@0,	1992-07-21,	1992-07-21  ]
[ 46@0,	1992-07-02,	1992-07-02  ]
[ 47@0,	1992-07-10,	1992-07-10  ]
[ 48@0,	1996-09-29,	1996-09-29  ]
[ 49@0,	1996-11-14,	1996-11-14  ]
[ 50@0,	1996-11-04,	1996-11-04	]
[ 51@0,	1996-09-26,	1996-09-26  ]
[ 52@0,	1996-10-02,	1996-10-02  ]
[ 53@0,	1996-10-17,	1996-10-17  ]
[ 54@0,	1996-12-08,	1996-12-08  ]
[ 55@0,	1994-09-30,	1994-09-30  ]
[ 56@0,	1995-04-20,	1995-04-20  ]
[ 57@0,	1995-07-17,	1995-07-17  ]
[ 58@0,	1995-07-06,	1995-07-06  ]
[ 59@0,	1994-02-19,	1994-02-19	]
[ 60@0,	1994-02-21,	1994-02-21  ]
[ 61@0,	1997-04-17,	1997-04-17  ]
[ 62@0,	1997-01-27,	1997-01-27  ]
[ 63@0,	1997-02-20,	1997-02-20  ]
[ 64@0,	1997-03-18,	1997-03-18  ]
[ 65@0,	1997-04-19,	1997-04-19  ]
[ 66@0,	1997-01-25,	1997-01-25  ]
[ 67@0,	1998-07-04,	1998-07-04  ]
[ 68@0,	1998-06-26,	1998-06-26	]
[ 69@0,	1998-08-13,	1998-08-13  ]
[ 70@0,	1998-06-27,	1998-06-27  ]
[ 71@0,	1998-06-19,	1998-06-19  ]
[ 72@0,	1998-08-11,	1998-08-11  ]
[ 73@0,	1998-06-24,	1998-06-24	]
[ 74@0,	1994-08-17,	1994-08-17  ]
[ 75@0,	1994-08-24,	1994-08-24  ]
[ 76@0,	1994-07-02,	1994-07-02  ]
[ 77@0,	1994-06-06,	1994-06-06  ]
[ 78@0,	1994-07-31,	1994-07-31  ]
[ 79@0,	1994-10-03,	1994-10-03  ]
[ 80@0,	1994-01-12,	1994-01-12  ]
[ 81@0,	1994-03-03,	1994-03-03  ]
[ 82@0,	1994-01-26,	1994-01-26  ]
[ 83@0,	1994-03-17,	1994-03-17  ]
[ 84@0,	1994-02-13,	1994-02-13  ]
[ 85@0,	1994-01-26,	1994-01-26  ]
[ 86@0,	1998-04-10,	1998-04-10  ]
[ 87@0,	1998-05-23,	1998-05-23  ]
[ 88@0,	1998-02-23,	1998-02-23  ]
[ 89@0,	1998-04-12,	1998-04-12  ]
[ 90@0,	1998-01-29,	1998-01-29  ]
[ 91@0,	1998-03-05,	1998-03-05	]
[ 92@0,	1994-07-19,	1994-07-19  ]
[ 93@0,	1994-06-03,	1994-06-03  ]
[ 94@0,	1993-04-01,	1993-04-01  ]
[ 95@0,	1993-04-13,	1993-04-13  ]
[ 96@0,	1993-05-14,	1993-05-14  ]
[ 97@0,	1994-12-24,	1994-12-24  ]
[ 98@0,	1994-12-01,	1994-12-01  ]
[ 99@0,	1994-12-30,	1994-12-30  ]

# 07:42:13 >  
# 07:42:13 >  "Done."
# 07:42:13 >  

